imagebuildah/stage_executor.go

```diff
@@ -1168,6 +1193,24 @@ func (s *StageExecutor) Execute(ctx context.Context, base string) (imgID string,
 			if err != nil {
 				return "", nil, fmt.Errorf("error checking if cached image exists from a previous build: %w", err)
 			}
+			// All the best effort to find cache on localstorage have failed try pulling
+			// cache from remote repo if `--cache-from` was configured.
+			if cacheID == "" && s.executor.cacheFrom != "" {
+				src := fmt.Sprintf("%s:%s", s.executor.cacheFrom, cacheKey)
+				// only attempt to use cache again if pulling was successful
+				// otherwise do nothing and attempt to run the step, err != nil
+				// is ignored and will be automatically logged for --log-level debug
+				if id, err := s.pullCache(ctx, src); id != "" && err == nil {
+					logCachePulled(src)
+					cacheID, err = s.intermediateImageExists(ctx, node, addedContentSummary, s.stepRequiresLayer(step))
```

Comment:
BTW it might be worthwhile to extract the “does $candidateImage match the current image and next step” logic from `intermediateImageExists` into a separate function, and here call it _only_ with the just-pulled image.

`intermediateImageExists` seems to do several a few c/storage operations (at least the `.Layer` call is not cached), each with its own locking syscalls and related overhead, so doing an O(1) check instead of an O(existing images) check would be nice.

Reply:
(Refactorings like this would be best done in separate commits, or maybe deferred to a separate PR.)

Reply:
Will tackle this in a different PR.